utils/queuer.go

```diff
@@ -50,12 +58,33 @@ func (q GeneralKVQueue) Enqueue(key Key, value codec.ProtoMarshaler) {
 	q.store.Set(key, value)
 }
 
-// Dequeue pops the bottom of the queue and unmarshals it into the given object, and return true if anything
-// in the queue is found and the value passes the optional filter function
-func (q GeneralKVQueue) Dequeue(value codec.ProtoMarshaler, filter ...func(value codec.ProtoMarshaler) bool) bool {
+func defaultFilter(_ codec.ProtoMarshaler) (pass bool, stop bool) { return true, false }
+
+// Dequeue pops the first item in queue and stores it in the given value
+func (q GeneralKVQueue) Dequeue(value codec.ProtoMarshaler) bool {
 	iter := sdk.KVStorePrefixIterator(q.store.KVStore, q.name.AsKey())
 	defer CloseLogError(iter, q.logger)
 
+	return q.dequeue(value, iter, defaultFilter)
+}
+
+// DequeueIf pops the first item in queue iff it matches the given filter and stores it in the given value
+func (q GeneralKVQueue) DequeueIf(value codec.ProtoMarshaler, filter func(value codec.ProtoMarshaler) bool) bool {
+	iter := sdk.KVStorePrefixIterator(q.store.KVStore, q.name.AsKey())
+	defer CloseLogError(iter, q.logger)
+
+	return q.dequeue(value, iter, func(value codec.ProtoMarshaler) (bool, bool) { return filter(value), false })
+}
+
+// DequeueUntil pops the first item in queue that matches the given filter and stores it in the given value
+func (q GeneralKVQueue) DequeueUntil(value codec.ProtoMarshaler, filter func(value codec.ProtoMarshaler) bool) bool {
+	iter := sdk.KVStorePrefixIterator(q.store.KVStore, q.name.AsKey())
+	defer CloseLogError(iter, q.logger)
+
+	return q.dequeue(value, iter, func(value codec.ProtoMarshaler) (bool, bool) { return filter(value), true })
```

Comment:
After splitting out the `Dequeue`s, I'm not sure if `filter` needs to be wrapped. Could just pass `continueIfNotQualified` as a separate arg to `dequeue` since it's independent of `filter`, unless you think this might be useful later?

Reply:
As an implementation detail, this is the most generic way of doing it. I like keeping possibilities open for any future.